### PR TITLE
Feature/372 Export CT cube

### PR DESCRIPTION
**Pull request by Yoanka**
Work in progress, so far Voxelplan export works correctly.

### CODE REVIEW

**Comment by Yoanka** [comment]
DICOM export relies on the following PyTRiP changes:
https://github.com/pytrip/pytrip/pull/586

File: pytripgui/app_logic/app_callbacks.py [hunk 1]
```diff
@@ -216,6 +219,41 @@ def open_dicom_callback(self, patient_item):
...
+        path = self.parent_gui.save_file_path("Export patient to Voxelplan", "Voxelplan (*.hed)")
+
+        if path == "":
+            return False
+
+        path_base, extension = os.path.splitext(path)
+        path, basename = os.path.split(path_base)
+        logger.info("Voxelplan export to: " + path + " with plan basename: " + basename)
+
+        patient_item.data.ctx.write(os.path.join(path, basename + ".ctx"))
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:234]: Use extension stored in CtxCube class https://github.com/pytrip/pytrip/blob/master/pytrip/ctx.py#L40:
```
class CtxCube(Cube):
    """ Class for handling CT-data. In TRiP98 these are stored in VOXELPLAN format with the .ctx suffix.
    This class can also handle Dicom files.
    """
    data_file_extension = '.ctx'
```

File: pytripgui/app_logic/app_callbacks.py [hunk 2]
```diff
@@ -216,6 +219,41 @@ def open_dicom_callback(self, patient_item):
...
+        patient_item.data.vdx.write(os.path.join(path, basename + ".vdx"))
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:235]: Same as above:
https://github.com/pytrip/pytrip/blob/master/pytrip/vdx.py#L93

Comment (grzanka) [reply pytripgui/app_logic/app_callbacks.py:235]: What will happen if VDX data is not present ?

File: pytripgui/app_logic/app_callbacks.py [hunk 3]
```diff
@@ -216,6 +219,41 @@ def open_dicom_callback(self, patient_item):
...
+        logger.debug("DICOM export start.")
+        path = self.parent_gui.browse_folder_path("Export patient to DICOM")
+
+        if path == "":
+            return False
+
+        logger.info("DICOM export to: " + path)
+
+        patient_item.data.ctx.write_dicom(path)
+        patient_item.data.vdx.write_dicom(path)
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:252]: What will happen if no vdx data is present ?

File: pytripgui/main_window_qt_vc/main_window_view.py [hunk 4]
```diff
@@ -38,6 +38,10 @@ def browse_file_path(self, name, extension, path=None):
         selected_file = QFileDialog.getOpenFileName(self.ui, name, path, extension)
         return selected_file[0]
 
+    def save_file_path(self, caption, extension, path=None):
+        file = QFileDialog.getSaveFileName(self.ui, caption, path, extension)
+        return file[0]
```
Comment (grzanka) [inline pytripgui/main_window_qt_vc/main_window_view.py:43]: Add a comment explaining why `file[0]` is returned, not simply `file`
Change `file` to something more meaningful - is it path to the file, file object or a filename ?

File: pytripgui/tree_vc/tree_view.py [hunk 5]
```diff
@@ -57,6 +58,8 @@ def _internal_context_menu_callback(self, pos):
         elif isinstance(self.selected_item, PatientItem):
             popup_menu.addAction("Open Voxelplan", self.internal_events.on_open_voxelplan)
             popup_menu.addAction("Add new Plan", self.internal_events.on_add_child)
+            popup_menu.addAction("Export patient as Voxelplan", self.internal_events.on_export_patient_voxelplan)
```
Comment (grzanka) [inline pytripgui/tree_vc/tree_view.py:61]: Rename to "Export as Voxelplan"

File: pytripgui/tree_vc/tree_view.py [hunk 6]
```diff
@@ -57,6 +58,8 @@ def _internal_context_menu_callback(self, pos):
...
+            popup_menu.addAction("Export patient as Dicom", self.internal_events.on_export_patient_dicom)
```
Comment (grzanka) [inline pytripgui/tree_vc/tree_view.py:62]: Rename to "Export as DICOM"

**Review by grzanka — CHANGES_REQUESTED** [review]

File: pytripgui/tree_vc/tree_view.py (hunk 6 above)
Comment (Yoanka) [reply pytripgui/tree_vc/tree_view.py]: done

File: pytripgui/tree_vc/tree_view.py (hunk 5 above)
Comment (Yoanka) [reply pytripgui/tree_vc/tree_view.py]: done

File: pytripgui/main_window_qt_vc/main_window_view.py (hunk 4 above)
Comment (Yoanka) [reply pytripgui/main_window_qt_vc/main_window_view.py]: Modified the name to selected_file_path and named the other member of the returned tuple.

File: pytripgui/app_logic/app_callbacks.py (hunk 1 above)
Comment (Yoanka) [reply pytripgui/app_logic/app_callbacks.py]: done

File: pytripgui/tree_vc/tree_view.py [hunk 7]
```diff
@@ -57,6 +58,8 @@ def _internal_context_menu_callback(self, pos):
         elif isinstance(self.selected_item, PatientItem):
             popup_menu.addAction("Open Voxelplan", self.internal_events.on_open_voxelplan)
             popup_menu.addAction("Add new Plan", self.internal_events.on_add_child)
+            popup_menu.addAction("Export as Voxelplan", self.internal_events.on_export_patient_voxelplan)
+            popup_menu.addAction("Export as Dicom", self.internal_events.on_export_patient_dicom)
```
Comment (grzanka) [inline pytripgui/tree_vc/tree_view.py:62]: DICOM, not Dicom

File: pytripgui/main_window_qt_vc/main_window_view.py [hunk 8]
```diff
@@ -35,8 +35,15 @@ def browse_file_path(self, name, extension, path=None):
...
         :return full file path, or empty string
         """
-        selected_file = QFileDialog.getOpenFileName(self.ui, name, path, extension)
-        return selected_file[0]
+        selected_file_path, selected_file_extension = QFileDialog.getOpenFileName(self.ui, name, path, extension)
+        return selected_file_path
+
+    def save_file_path(self, caption, extension, path=None):
+        """
+        :return full file, or empty string
```
Comment (grzanka) [inline pytripgui/main_window_qt_vc/main_window_view.py:42]: What is full file ? File full of bytes or a path ? Is it absolute path, relative path ?

Comment (Yoanka) [reply pytripgui/main_window_qt_vc/main_window_view.py]: changed to "full file path (absolute)"

File: pytripgui/tree_vc/tree_view.py (hunk 7 above)
Comment (Yoanka) [reply pytripgui/tree_vc/tree_view.py]: fixed

File: pytripgui/app_logic/app_callbacks.py (hunk 2 above)
Comment (Yoanka) [reply pytripgui/app_logic/app_callbacks.py]: VDX import is now skipped if it doesn't exist.

File: pytripgui/app_logic/app_callbacks.py (hunk 3 above)
Comment (Yoanka) [reply pytripgui/app_logic/app_callbacks.py:252]: VDX import is now skipped if it doesn't exist.

File: pytripgui/app_logic/app_callbacks.py [hunk 9]
```diff
@@ -216,6 +219,62 @@ def open_dicom_callback(self, patient_item):
...
+        Parameters:
+        patient_item (PatientItem): Patient tree item containing the patient's data
+
+        Returns:
+        bool: Whether export was successful
+        """
+        logger.debug("Voxelplan export start.")
+        path = self.parent_gui.save_file_path("Export patient to Voxelplan", "Voxelplan (*.hed)")
+
+        if path == "":
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:235]: More pythonic would be:
```
if not path:
   return False
```

File: pytripgui/app_logic/app_callbacks.py [hunk 10]
```diff
@@ -216,6 +219,62 @@ def open_dicom_callback(self, patient_item):
...
+            return False
+
+        path_base, extension = os.path.splitext(path)
+        path, basename = os.path.split(path_base)
+        logger.info("Voxelplan export to: " + path + " with plan basename: " + basename)
+
+        patient_item.data.ctx.write(os.path.join(path, basename + patient_item.data.ctx.data_file_extension))
+        if patient_item.data.vdx is not None:
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:243]: More pythonic would be:
```
if patient_item.data.vdx:
```

File: pytripgui/main_window_qt_vc/main_window_view.py [hunk 11]
```diff
@@ -33,20 +32,27 @@ def get_kernel_config_view():
...
+        :return full file path (absolute), or empty string
         """
-        selected_file = QFileDialog.getOpenFileName(self.ui, name, path, extension)
-        return selected_file[0]
+        selected_file_path, selected_file_extension = QFileDialog.getOpenFileName(self.ui, name, path, extension)
+        return selected_file_path
+
+    def save_file_path(self, caption, extension, path=None):
+        """
+        :return full file path (absolute), or empty string
```
[the remark this answers is not in the document]
Comment (grzanka) [reply pytripgui/main_window_qt_vc/main_window_view.py:42]: Explain under what circumstances this method returns empty string

File: pytripgui/main_window_qt_vc/main_window_view.py [hunk 12]
```diff
@@ -33,20 +32,27 @@ def get_kernel_config_view():
...
         """
-        :return full file path, or empty string
+        :return full directory path (absolute), or empty string
         """
         dialog = QFileDialog(self.ui, name, path)
         dialog.setFileMode(QFileDialog.Directory)
         dialog.setOptions(QFileDialog.ShowDirsOnly)
         dialog.exec_()
-        selected_path = os.path.join(dialog.directory().path(), '')
+        selected_path = dialog.selectedFiles()[0]
```
Comment (grzanka) [inline pytripgui/main_window_qt_vc/main_window_view.py:55]: Explain why `[0]` needs to be used here

**Review by grzanka — CHANGES_REQUESTED** [review]
more comments added

File: pytripgui/app_logic/app_callbacks.py (hunk 9 above)
Comment (Yoanka) [reply pytripgui/app_logic/app_callbacks.py]: changed

File: pytripgui/app_logic/app_callbacks.py (hunk 10 above)
Comment (Yoanka) [reply pytripgui/app_logic/app_callbacks.py]: changed

File: pytripgui/main_window_qt_vc/main_window_view.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Yoanka) [reply pytripgui/main_window_qt_vc/main_window_view.py]: I improved the docstrings of the 3 file browsing functions, so hopefully now they describe them properly.

File: pytripgui/main_window_qt_vc/main_window_view.py (hunk 12 above)
Comment (Yoanka) [reply pytripgui/main_window_qt_vc/main_window_view.py:64]: Comment added. The only method that extracts selected files from a file dialog window returns a list, even though this window only allows you to select a single file.
https://github.com/pytrip/pytripgui/blob/284f696219f254469a8a8ea5b5407463d711d657/pytripgui/main_window_qt_vc/main_window_view.py#L62

**Review by grzanka — APPROVED** [review]
Looks OK

**Comment by grzanka** [comment]
Exporting CT cube may takes significant time. User is not aware when it will be completed. Consider adding a progress bar here.
Maybe some modifications on pytrip side will be needed, to notify on the conversion progress.

**Pull request merged**